pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -192,6 +213,9 @@ func (h *upgradeHandler) OnChanged(_ string, upgrade *harvesterv1.Upgrade) (*har
 
 	// clean upgrade repo VMs and images if a upgrade succeeds.
 	if harvesterv1.UpgradeCompleted.IsTrue(upgrade) {
+		// try to restore vm to the state before upgrade
+		h.restoreVMState(upgrade)
+
 		// try to clean up images before purging the repo VM
 		_, exists := upgrade.Annotations[imageCleanupPlanCompletedAnnotation]
 		if exists {
```

Comment:
Just a suggestion: Could we move the `restoreVMState` call after the image-cleanup annotation check? That way, the function would only be called once. It is called twice currently, and there will be a "Get secret failed" message logged.

Reply:
Sure, thank you.